# Mann-Whitney U Testi
Mann-Whitney U Testi, Bağımsız İki Örneklem T Testi'nin (Konu 59) 
normallik varsayımı sağlanmadığında kullanılan non-parametrik 
karşılığıdır. Bu, Non-Parametrik Testler ailesinde **en sık kullanılan** 
testlerden biri.

## Ne Zaman Kullanılır?
İki **bağımsız** grubun dağılımlarını karşılaştırmak istediğinde, ama 
normallik sağlanmadığında (ya da örneklem çok küçükse, ya da veri 
ordinal/sıralı ise).

## Mann-Whitney U Testinin Mantığı
İşaret Testi'nden farklı olarak (sadece yöne bakıyordu), Mann-Whitney 
**tüm gözlemleri BİRLEŞTİRİP sıralar** (rank verir), sonra her grubun 
sıralamalardaki **toplam rütbesine** bakar. Eğer bir grup sistematik 
olarak diğerinden **daha yüksek sıralarda** yer alıyorsa, bu, o grubun 
değerlerinin genel olarak daha büyük olduğuna işaret eder.

## Hipotezler
- **H0:** İki grubun dağılımları aynıdır (birinden rastgele çekilen bir değerin, diğerinden çekilen bir değerden büyük olma olasılığı %50'dir)
- **H1:** İki grubun dağılımları farklıdır

## T Testinden Farkı
T testi **ortalamaları** karşılaştırırken, Mann-Whitney **dağılımların 
genel konumunu/sıralamasını** karşılaştırır — bu yüzden sonuçları "medyan 
farkı" gibi yorumlamak daha doğrudur, "ortalama farkı" değil.

## Python'da Kullanımı
```python
from scipy.stats import mannwhitneyu

istatistik, p_degeri = mannwhitneyu(grup1, grup2, alternative='two-sided')
```

## Etki Büyüklüğü
Mann-Whitney için de bir etki büyüklüğü hesaplanabilir (rank-biserial 
correlation) Cohen's d'nin (Konu 52) sıralamaya dayalı versiyonu.

In [2]:
# Senaryo: İki Farklı Ürün Sayfası Tasarımından Gelen Ziyaretçilerin Sitede Kalma Süresi
import numpy as np
from scipy import stats
np.random.seed(42)

tasarim_A = np.random.exponential(scale=45, size=25)
tasarim_B = np.random.exponential(scale=65, size=25)

# Öncelikle parametrik veya parametrik olmayan testlerden hangisini uygulayacağımıza yönelik normallik testi yapalım.
# H0: Veriler normal dağılıma uyar
# H1: Veriler normal dağılıma uymaz
shapiro_A_pvalue = stats.shapiro(x=tasarim_A).pvalue # type: ignore
shapiro_B_pvalue = stats.shapiro(x=tasarim_B).pvalue # type: ignore
alpha = 0.05
print(f'Tasarım A"ya ait Shapiro-Wilk testinin P Değeri: {shapiro_A_pvalue}')
print(f'Tasarım B"ye ait Shapiro-Wilk testinin P Değeri: {shapiro_B_pvalue}')

Tasarım A"ya ait Shapiro-Wilk testinin P Değeri: 0.00010947058891936719
Tasarım B"ye ait Shapiro-Wilk testinin P Değeri: 0.00035691879347353754


Her iki Tasarım Grubuna ait olan veriler normal dağılmadığı için bağımsız örneklem *T*-testinin alternafi olan Mann-Whitney U testini kullanacağız. 

In [5]:
# Mann-Whitney U testine ait Hipotez yapısı
# H0: İki grubun dağılımları aynıdır (birinden rastgele çekilen bir değerin, diğerinden çekilen bir değerden büyük olma olasılığı %50'dir).
# H1: İki grubun dağılımları farklıdır
mnu_statistic, mnu_pvalue = stats.mannwhitneyu(
	x=tasarim_A,
	y=tasarim_B,
	alternative='two-sided'
)
if mnu_pvalue < alpha:
    print('H0 hipotezini reddecek yeterli kanıta sahibiz, iki grubun dağılımları anlamlı biçimde birbirinden farklı olabilir.')
else:
    print('H0 hipotezini reddecek yeterli kanıta sahip değiliz, iki grubun dağılımları aynıdır.')

H0 hipotezini reddecek yeterli kanıta sahip değiliz, iki grubun dağılımları aynıdır.


In [7]:
print(f'Mann-Whitney U testi P-Value: {mnu_pvalue}')
print(f'Mann-Whitney U testi U-Statistic: {mnu_statistic}')

Mann-Whitney U testi P-Value: 0.2859010589675479
Mann-Whitney U testi U-Statistic: 257.0


In [6]:
tasarim_A_ortalama, tasarim_A_medyan, tasarim_B_ortalama, tasarim_B_medyan = np.mean(tasarim_A), np.median(tasarim_A), np.mean(tasarim_B), np.median(tasarim_B)
print(f'A tasarımının ortalaması: {tasarim_A_ortalama}')
print(f'A tasarımının medyanı: {tasarim_A_medyan}')
print(f'B tasarımının ortalaması: {tasarim_B_ortalama}')
print(f'B tasarımının medyanı: {tasarim_B_medyan}')

A tasarımının ortalaması: 37.343231248899556
A tasarımının medyanı: 21.11706404895866
B tasarımının ortalaması: 56.03255791533103
B tasarımının medyanı: 44.43056982871976


### Sonuç
İki tasarımın sitede kalma süresi dağılımlarının farklı olup olmadığını 
test etmek amacıyla Mann-Whitney U Testi uyguladık (normallik 
sağlanmadığı için, p<0.001 her iki grupta da). Test sonuçlarına göre, 
gözlenen betimsel fark (Tasarım A medyan=21.1, Tasarım B medyan=44.4) 
istatistiksel olarak anlamlı bulunmamıştır (U=257.0, p=0.286). Bu 
sonuç, gerçekte bir fark olmadığını gösterebileceği gibi, mevcut 
örneklem büyüklüğünün (n=25 her grup) bu farkı yakalamak için yetersiz 
kalmış olabileceğine de işaret edebilir, bu nedenle daha büyük bir 
örneklemle testin tekrarlanması önerilir.